mesonbuild/interpreter/interpreter.py

```diff
@@ -2215,67 +2214,68 @@ def func_install_subdir(self, node: mparser.BaseNode, args: T.Tuple[str],
         self.build.install_dirs.append(idir)
         return idir
 
-    @FeatureNewKwargs('configure_file', '0.47.0', ['copy', 'output_format', 'install_mode', 'encoding'])
-    @FeatureNewKwargs('configure_file', '0.46.0', ['format'])
-    @FeatureNewKwargs('configure_file', '0.41.0', ['capture'])
-    @FeatureNewKwargs('configure_file', '0.50.0', ['install'])
-    @FeatureNewKwargs('configure_file', '0.52.0', ['depfile'])
-    @FeatureNewKwargs('configure_file', '0.60.0', ['install_tag'])
-    @permittedKwargs({'input', 'output', 'configuration', 'command', 'copy', 'depfile',
-                      'install_dir', 'install_mode', 'capture', 'install', 'format',
-                      'output_format', 'encoding', 'install_tag'})
     @noPosargs
-    def func_configure_file(self, node, args, kwargs):
-        if 'output' not in kwargs:
-            raise InterpreterException('Required keyword argument "output" not defined.')
-        actions = {'configuration', 'command', 'copy'}.intersection(kwargs.keys())
-        if len(actions) == 0:
+    @typed_kwargs(
+        'configure_file',
+        DEPFILE_KW.evolve(since='0.52.0'),
+        INSTALL_MODE_KW.evolve(since='0.47.0,'),
+        INSTALL_TAG_KW.evolve(since='0.60.0'),
+        KwargInfo('capture', bool, default=False, since='0.41.0'),
+        KwargInfo(
+            'command',
+            (ContainerTypeInfo(list, (build.Executable, ExternalProgram, compilers.Compiler, mesonlib.File, str), allow_empty=False), NoneType),
+            listify=True,
+        ),
+        KwargInfo(
+            'configuration',
+            (ContainerTypeInfo(dict, (str, int, bool)), build.ConfigurationData, NoneType),
+        ),
+        KwargInfo('copy', bool, default=False, since='0.47.0'),
+        KwargInfo('encoding', str, default='utf-8', since='0.47.0'),
+        KwargInfo('format', str, default='meson', since='0.46.0',
+                  validator=in_set_validator({'meson', 'cmake', 'cmake@'})),
+        KwargInfo(
+            'input',
+            ContainerTypeInfo(list, (mesonlib.File, str)),
+            listify=True,
+            default=[],
+        ),
+        # Cannot use shared implementation until None backwards compat is dropped
+        KwargInfo('install', (bool, NoneType), since='0.50.0'),
+        KwargInfo('install_dir', (str, bool), default='',
+                  validator=lambda x: 'must be `false` if boolean' if x is True else None),
+        KwargInfo('output', str, required=True),
+        KwargInfo('output_format', str, default='c', since='0.47.0',
+                  validator=in_set_validator({'c', 'nasm'})),
+    )
+    def func_configure_file(self, node: mparser.BaseNode, args: T.List[TYPE_var],
+                            kwargs: kwargs.ConfigureFile):
+        actions = sorted(x for x in {'configuration', 'command'} if kwargs[x] is not None)
+        if kwargs['copy']:
+            actions.append('copy')
```

Comment:
Hmm, perhaps `if kwargs[x] not in {None, False}`?

Reply:
I guess yeah, we could do that. I don't really know why we'd care to check `is None` in the `elif` case above, we can never get a falsy value that isn't None since we don't allow `command` to be empty